finp2p-contracts/contracts/token/ERC20/FINP2POperatorERC20.sol

```diff
@@ -174,11 +174,27 @@ contract FINP2POperatorERC20 is IFinP2PAsset, IFinP2PEscrow, AccessControl {
 
         Asset memory asset = assets[assetId];
         uint256 balance = IERC20(asset.tokenAddress).balanceOf(issuer);
-        require(balance >= quantity, "Not sufficient balance to redeem");
-
-        ERC20WithOperator(asset.tokenAddress).burnFrom(issuer, quantity);
-
-        emit Redeem(assetId, account, quantity);
+        if (operationId == 0x00000000000000000000000000000000) {
+           require(balance >= quantity, "Not sufficient balance to redeem");
+        }
+        if (operationId != 0x00000000000000000000000000000000) {
+                require(haveContract(operationId), "Contract does not exists");
+
+                Lock storage lock = locks[operationId];
+                require(quantity == lock.amount, "Amount to redeem is not equal to locked amount for this operationId");
```

Comment:
Shouldn't we also validate the balance is enough?

Reply:
the balance itself may not be enough, since we are transferring the held amount to the address of the contract on the hold action. As discussed I did add a check that there is enough balance on the contact address.